include/exadg/incompressible_navier_stokes/spatial_discretization/calculators/shear_rate_calculator.cpp

```diff
@@ -0,0 +1,93 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#include <exadg/incompressible_navier_stokes/spatial_discretization/calculators/shear_rate_calculator.h>
+
+namespace ExaDG
+{
+namespace IncNS
+{
+template<int dim, typename Number>
+ShearRateCalculator<dim, Number>::ShearRateCalculator()
+  : matrix_free(nullptr), dof_index_u(0), dof_index_u_scalar(0), quad_index(0)
+{
+}
+
+template<int dim, typename Number>
+void
+ShearRateCalculator<dim, Number>::initialize(dealii::MatrixFree<dim, Number> const & matrix_free_in,
+                                             unsigned int const                      dof_index_u_in,
+                                             unsigned int const dof_index_u_scalar_in,
+                                             unsigned int const quad_index_in)
+{
+  matrix_free        = &matrix_free_in;
+  dof_index_u        = dof_index_u_in;
+  dof_index_u_scalar = dof_index_u_scalar_in;
+  quad_index         = quad_index_in;
+}
+
+template<int dim, typename Number>
+void
+ShearRateCalculator<dim, Number>::compute_shear_rate(VectorType & dst, VectorType const & src) const
+{
+  dst = 0;
+
+  matrix_free->cell_loop(&This::cell_loop, this, dst, src);
+}
+
+template<int dim, typename Number>
+void
+ShearRateCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> const & matrix_free,
+                                            VectorType &                            dst,
+                                            VectorType const &                      src,
+                                            Range const & cell_range) const
+{
+  CellIntegratorVector integrator_vector(matrix_free, dof_index_u, quad_index);
+  CellIntegratorScalar integrator_scalar(matrix_free, dof_index_u_scalar, quad_index);
+
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+  {
+    integrator_vector.reinit(cell);
+    integrator_vector.gather_evaluate(src, dealii::EvaluationFlags::gradients);
+
+    integrator_scalar.reinit(cell);
+
+    for(unsigned int q = 0; q < integrator_scalar.n_q_points; q++)
+    {
+      dealii::SymmetricTensor<2, dim, dealii::VectorizedArray<Number>> sym_grad_u =
+        integrator_vector.get_symmetric_gradient(q);
+      scalar shear_rate = std::sqrt(0.5 * scalar_product(sym_grad_u, sym_grad_u));
```

Comment:
Is there a common definition of how to define the shear rate? I remember the implementation of the Smagorinsky turbulence model where we have the factor of 2 instead of 0.5, see https://github.com/exadg/exadg/blob/ef064e0a5651ad46cee8205cfeccb203960443e5/include/exadg/incompressible_navier_stokes/spatial_discretization/turbulence_model.cpp#L282.

In case there are conventions involved, it could make sense to mention them and potentially give a reference to literature.

Reply:
Thanks for pointing that out, in fact, definitions with 0.5 and 2 are being used ... which is a mess.
I have been using 0.5 in the past (changing the constitutive parameters accordingly), BUT using 2 instead is (1) the one consistent with 1D Poiseuille flow, (2) more common (which is debatable?), (3) coincides with the code snippet you mention, and we have a good source to cite:
https://doi.org/10.1007/978-3-7643-7806-6
So, I will swap out the 0.5 for 2.0.